File: crc_8_16/CRC_3/crc8.c
#include "crc8.h"

static unsigned char crctab[8];

/*
 * Generates an unsigned char table with 8 entries.
 */
void crc8_generate_table(unsigned char poly)
{
	int i;
	unsigned char residue = poly;

	for (i=0;i<8;i++) {
		int carry;

		crctab[i] = residue;
		carry = residue & 0x80;
		residue <<= 1;
		if (carry)
			residue ^= poly;
	}
}

/*
 * Single iteration of a the CRC computation
 */
unsigned char crc8_single(unsigned char a, unsigned char b)
{
	unsigned char *residue = crctab;

	while (b) {
		if (b & 1)
			a ^= *residue;
		b >>= 1;
		residue++;
	}

	return a;
}

/*
 * Compute CRC8 of a string
 */
unsigned char crc8_compute(unsigned char init, const unsigned char *data, int count)
{
	unsigned char crc = init;

	while (count--) {
		crc = crc8_single(*data, crc);
		data++;
	}

	return crc;
}

File: crc_8_16/CRC_3/crc8.py
class CRC8(object):
    # Precompute the table
    def __init__(self, poly, init):
        self.crctab = []
        self.init = init

        current_residue = poly
        for i in range(0, 8):
            self.crctab.append(current_residue)
            current_residue <<= 1
            if (current_residue & 0x100):
                current_residue = (current_residue & 0xff) ^ poly

    # Single iteration
    def single(self, left, right):
        mask = 1
        for residue in self.crctab:
            if right & mask:
                left ^= residue
            mask <<= 1
        return left

    # Check if the packet's CRC8 is valid
    def check(self, bytelist):
        crc = self.compute(bytelist)
        if crc:
            return False
        return True

    # Compute the CRC8 of a list of bytes
    def compute(self, bytelist):
        current_crc = self.init
        for byte in bytelist:
            current_crc = self.single(byte, current_crc)
        return current_crc


